### DISCUSSION TITLE
Pocketbase as a framework on Pockethost

### DISCUSSION THREAD

**Discussion by dindinet** [pockethost/pockethost#192] [Q&A] [answered]
Hi 

Is it posible to host on pockethost using a golang app that uses pocketbase as a Framework?

**Comment by benallfree** [comment, marked as answer]
No. PocketHost runs PocketBase instances in a shared environment where custom builds would pose security threats. If we ever isolate PocketBase instances, this could be possible. 

Can you share what customizations you’re trying to use? There may be other solutions.

**Reply by dindinet** [reply]
Thanks benallfree I thought as much but needed to ask.

the customisations I envisage so far
OnRecordBeforeUpdateRequest fetch data from Cloudinary Api to include image urls in the datawrite
OnRecordAfterUpdateRequest where I want to send an email after successful data write of enq form

**Reply by benallfree** [reply]
Got it. 

It’s not well documented yet, but both of those things can be accomplished with our Deno cloud workers instead of custom PB. The Deno workers are written in Typescript and run in a Deno sandbox alongside the PB instance. You can do almost anything you want in your Deno worker, including using the PB js-sdk to listen for PB changes, performing interval checks, sending emails, etc. 

In the Deno worker, you can use the PB realtime API to listen for changes and respond. 

In the case of doing an API fetch, OnRecordBeforeUpdateRequest is not the best idea because then you’re blocking a record update with something that could fail or take a long time. Instead, write the record in an intermediate/incomplete form and use a realtime listener to perform the API lookup and record update out-of-band. 

Same with OnRecordAfterUpdateRequest. Since sending emails can fail and/or need to be retried, it is better to keep a queue of pending notifications and process them out of band. Add a status column (or collection) to track the state of each notification.  

For all of the above reasons, I much prefer to lock my collections so they’re writable only by admin accounts (which the Deno worker uses). Then I create an “rpc” collection where I can insert untrusted rpc calls from the front end. The Deno worker will then validate the request and create PB records on behalf of the client. That way my data stays in a known state and untrusted clients don’t have the ability to write garbage to my DB. Since the Deno worker listens for rpc inserts, it also has the ability to filter/throttle requests using whatever criteria I decide. 

I’ll need to document that pattern more too.

**Comment by dindinet** [comment]
Hi

Appreciate your suggestions. I will explore deno workers.

The use case is for a backend used by a very small (1-5) team as a
headless CMS. The content will be used to populate a static site deployed
to Netlify.

I wanted to stick to the pocketbase ethos of 'Open Source Backend ... in
one file'

I tested the api call for OnRecordBeforeUpdateRequest and the
combined submitted and fetched data were in the sqlite database before I
could blink.
Similarly with sending mail. I use mailersend.com which works great with
PB. The use case is for the sales team to get a heads up when a web form is
submitted.
Currently we do not write form data to database just forward to team inbox.
PB makes it so easy to add form data to database It seems mean not to do
it. :)

**Comment by Klunk75** [comment]
Hi all,
@benallfree I really appreciate an example and the documentation about Deno Workers (the relateted doc page is still almost empty: https://pockethost.io/docs/cloud_functions), I need it to communicate via API HTTP requests to another server (for example OneSignal, for push notifications) upon changes in the database (create, update, delete)... and other server side useful things.
Thanks a lot for your work!

Paolo

**Reply by benallfree** [reply]
@Klunk75 At this time external HTTP requests are not supported, but this is an interesting idea. I'll have to think more about it!

**Comment by Sleepful** [comment]
Pockethost looks very promising, but it is difficult to see it as a better option than self-hosting my own Pocketbase instance if it has so many limitations. Just my 2 cents, very cool project, wish it had better docs on self-hosting too. If I could self-host pocketHOST then I would be able to use my own custom binaries. Seems like a lot of potential!

**Reply by benallfree** [reply]
Why aren't you using JS hooks?
